account_invoice_fiscal_position_update/models/account_invoice.py

```diff
@@ -27,13 +27,18 @@ def fiscal_position_change(self):
                     account = (
                         product.property_account_income_id or
                         product.categ_id.property_account_income_categ_id)
-                    taxes = product.taxes_id
+                    # M2M fields don't have an option 'company_dependent=True'
+                    # so we need per-company post-filtering
+                    taxes = product.taxes_id.filtered(
+                        lambda tax: tax.company_id == self.company_id)
                 else:
                     account = (
                         product.property_account_expense_id or
                         product.categ_id.property_account_expense_categ_id)
-                    taxes = product.supplier_taxes_id
-                taxes = taxes or account.tax_ids
+                    taxes = product.supplier_taxes_id.filtered(
```

Comment:
But if you think about it carefully, you'll see that doing it at the end will modify the behavior: if you have a tax in another company on a product, then
```
taxes = taxes or account.tax_ids
```
will copy taxes on taxes and then it will be filtered and the taxes will be removed. With the current code, it would have taken account.tax_ids.
I'm just saying that to underline the fact that moving the 3 filtered at the end won't give the same result.